### PR TITLE
santad: Log file changes, use prefix trees

### PR DESCRIPTION
**Pull request by russellhancox**
The way that the prefix trees are used and the giant switch statements around file write logging are awful, we need to refactor all of this in the near future.

### CODE REVIEW

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                                                 length:arg.length
+                                               encoding:NSUTF8StringEncoding]];
+      }
+      sm.args_array = (void *)CFBridgingRetain(args);
+
+      callback = self.logCallback;
+      audit_token = m->event.exec.target->audit_token;
+      sm.ppid = m->event.exec.target->original_ppid;
+
+      size_t l = m->event.exec.target->executable->path.length;
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: What if `path_truncated` is true? Same throughout.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+  if (self.prefixTree->HasPrefix(sm.path)) {
+    return;
+  }
+
+  if (callback) {
+    sm.uid = audit_token_to_ruid(audit_token);
+    sm.gid = audit_token_to_rgid(audit_token);
+    sm.pid = audit_token_to_pid(audit_token);
+    proc_name(sm.pid, sm.pname, 1024);
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: I think this gets truncated to MAXCOMLEN ~32 Bytes.
Does `[NSRunningApplication runningApplicationWithProcessIdentifier:sm.pid].localizedName` work at this stage?

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    case ACTION_RESPOND_ALLOW:
+      ret = es_respond_auth_result(self.client, (es_message_t *)sm.es_message,
+                                   ES_AUTH_RESULT_ALLOW, true);
+      break;
+    case ACTION_RESPOND_DENY:
+      ret = es_respond_auth_result(self.client, (es_message_t *)sm.es_message,
+                                   ES_AUTH_RESULT_DENY, false);
+      break;
+    default:
+      return ES_RESPOND_RESULT_ERR_INVALID_ARGUMENT;
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: So we don't leak accidentally leak sm.es_message.
ret = ES_RESPOND_RESULT_ERR_INVALID_ARGUMENT

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        exit(ret);
+      default:
+        LOGE(@"Unable to create es client: %d. Sleeping for a minute.", ret);
+        sleep(60);
+        continue;
+    }
+  }
+}
+
+- (void)messageHandler:(const es_message_t *)m API_AVAILABLE(macos(10.15)) {
```
Comment (tburgin) [inline Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Somewhere in here we need to check prefixTree.

Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: We do, line 202

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm (hunk 3 above)
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm (hunk 2 above)
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: That seems to be less reliable than proc_name and proc_name is what 'old' Santa used.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+  // TODO(bur/rah): Currently this class only subscribes to exec events. Move this code
+  // somewhere exec specific if other types of events are added to this client.
+  santa_message_t sm;
+
+  audit_token_t audit_token = {};
```
[the remark this answers is not in the document]
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        sleep(60);
+        continue;
+    }
+  }
+}
+
+- (void)messageHandler:(const es_message_t *)m API_AVAILABLE(macos(10.15)) {
+  // TODO(bur/rah): Currently this class only subscribes to exec events. Move this code
+  // somewhere exec specific if other types of events are added to this client.
+  santa_message_t sm;
```
[the remark this answers is not in the document]
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+      default:
+        LOGE(@"Unable to create es client: %d. Sleeping for a minute.", ret);
+        sleep(60);
+        continue;
+    }
+  }
+}
+
+- (void)messageHandler:(const es_message_t *)m API_AVAILABLE(macos(10.15)) {
+  // TODO(bur/rah): Currently this class only subscribes to exec events. Move this code
```
[the remark this answers is not in the document]
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+  self = [super init];
+  if (self) {
+    [self establishClient];
+    _prefixTree = new SNTPrefixTree();
+  }
+
+  return self;
+}
+
+- (void)dealloc {
```
[the remark this answers is not in the document]
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Done

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+- (void)messageHandler:(const es_message_t *)m API_AVAILABLE(macos(10.15)) {
+  // TODO(bur/rah): Currently this class only subscribes to exec events. Move this code
+  // somewhere exec specific if other types of events are added to this client.
+  santa_message_t sm;
+
+  audit_token_t audit_token = {};
+  void (^callback)(santa_message_t);
+
+  switch (m->event_type) {
+    case ES_EVENT_TYPE_AUTH_EXEC: {
```
[the remark this answers is not in the document]
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: There's actually not that much overlap right now between the two except those vnode fields. I think we need to design a better agnostic format between santa_message_t and es_message_t that we can pass around to things like logging but that can come in a future PR.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm (hunk 1 above)
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: For logging I don't think it particularly matters but we can re-visit when we refactor this.

**Comment by russellhancox** [comment]
PTAL

**Comment by tburgin** [comment]
There is still a build error.
https://travis-ci.org/google/santa/builds/592772685#L704
I would like to see the prefix tree tests pass, given the lock changes.

File: Source/santad/EventProviders/SNTEndpointSecurityManager.mm (hunk 6 above)
[the remark this answers is not in the document]
Comment (russellhancox) [reply Source/santad/EventProviders/SNTEndpointSecurityManager.mm]: Reverted - putting a 0 in here causes errors, so I've made it {} here and in a few other places for consistency.

**Comment by russellhancox** [comment]
That's fixed now. Need to get rid of the Xcodeproj soon and go back to a single source of truth.

**Pull request merged**